### PR TITLE
[Merged by Bors] - fix: refactor of `fun_prop` and several bug fixes

### PR DESCRIPTION
**Pull request by lecopivo**
fix: refactor of `fun_prop` and several bug fixes

1. support for proving recursive functions
`fun_prop` correctly uses local hypothesis coming from induction step
2. unfolds let variables when necessary
3. unfolds `id`, `Function.comp`, `Function.uncurry`, `HasUncurry.uncurry`
Many theorems in mathlib are stated using `id`, `Function.comp` etc. to prevent restating all these theorems `fun_prop` just unfolds these common functions.
4. does not get blocked by `autoParam` or `outParam`
5. support for custom bundled morphism coercions. `DFunLike.coe` is not hardcoded anymore and any morphism coercion can be registered with `@[fun_prop_coe]` attribute
6. internal counter to prevent infinite loops
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/FunProp/Decl.lean [hunk 1]
```diff
@@ -28,20 +28,18 @@ structure FunPropDecl where
   /-- argument index of a function this function property talks about.
   For example, this would be 4 for `@Continuous α β _ _ f` -/
   funArgId : Nat
-  /-- Custom discharger for this function property. -/
-  dischargeStx? : Option (TSyntax `tactic)
   deriving Inhabited, BEq
 
-/-- -/
+/-- Disctriminatory tree for function properties. -/
```
Comment (j-loreaux) [inline Mathlib/Tactic/FunProp/Decl.lean:33]: Typo `Disct...`, but also, shouldn't this be a *discrimination* tree?
```suggestion
/-- Discrimination tree for function properties. -/
```

File: Mathlib/Tactic/FunProp/Decl.lean [hunk 2]
```diff
@@ -28,20 +28,18 @@ structure FunPropDecl where
...
 structure FunPropDecls where
-  /-- discriminatory tree for function properties -/
+  /-- Discriminatory tree for function properties. -/
```
Comment (j-loreaux) [inline Mathlib/Tactic/FunProp/Decl.lean:35]:
```suggestion
  /-- Discrimination tree for function properties. -/
```

File: Mathlib/Tactic/FunProp/Decl.lean [hunk 3]
```diff
@@ -28,20 +28,18 @@ structure FunPropDecl where
...
   decls : DiscrTree FunPropDecl := {}
   deriving Inhabited
 
 /-- -/
 abbrev FunPropDeclsExt := SimpleScopedEnvExtension FunPropDecl FunPropDecls
 
-/-- -/
+/-- Extension storing all function properties. -/
```
Comment (j-loreaux) [inline Mathlib/Tactic/FunProp/Decl.lean:42]: "all function properties" is a bit vague. Can we say:
```suggestion
/-- Extension storing function properties tracked and used by the `fun_prop` attribute and tactic, including, for example, `Continuous`, `Measurable`, `Differentiable`, etc. -/
```

**Comment by lecopivo** [comment]
Testing on sphere-eversion project reveals that things are still broken. I'm marking it as WIP for now.

File: Mathlib/Tactic/FunProp/MorExt.lean [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+
+private local instance : Ord Name := ⟨Name.quickCmp⟩
+
+
+/-- Environment extension storing all morphism coercions. -/
+initialize morCoeDeclsExt : SimpleScopedEnvExtension Name (Std.RBSet Name compare) ←
+  registerSimpleScopedEnvExtension {
+    name := by exact decl_name%
+    initial := .ofArray #[`DFunLike.coe] _
```
Comment (lecopivo) [inline Mathlib/Tactic/FunProp/MorExt.lean:27]: I'm a bit unsure about this. I need to add `@[fun_prop_coe]` attribute to `DFunLike.coe`. The question is do I hardcode it here or should I add early dependency of fun_prop to `Mathlib.Data.FunLike.Basic` and add the attribute there?

Comment (semorrison) [reply Mathlib/Tactic/FunProp/MorExt.lean:27]: I think I'd prefer it hardcoded here. You can add a comment.

**Comment by semorrison** [comment]
bors d+

File: Mathlib/Tactic/FunProp/MorExt.lean (hunk 4 above)
Comment (lecopivo) [reply Mathlib/Tactic/FunProp/MorExt.lean]: I followed Eric's recommendation and reused coercion extension already existing in std. So this issue disappeared

**Pull request closed without merging**